Use BTFGen

mauriciovasquezbernal opened the pull request:
BTFGen[0] allows to generate small BTF files with only the information
needed by a set of eBPF objects. This commit makes use of that tool
together with BTFHub[1] to create a set of BTF files needed by Inspektor
Gadget. These files are around 800KB and are included in the gadget
container. At container startup time, the entrypoint script checks if
there's a file for the running kernel, if yes it's installed on the
system.

[0]: https://github.com/kinvolk/btfgen
[1]: https://github.com/aquasecurity/btfhub

TODO:
- [x] Integrate with the CI to include this in the released container image  
- [x] Merge https://github.com/kinvolk/bcc/pull/15

mauriciovasquezbernal commented:
This PR is ready for review. Right now the whole BTFHub repo is cloned and the BTF files are generated each time the actions are run, this is something we need for improve for sure but I don't know how to do it now. I propose to review this PR in its current status and then we can think how to improve this process later on.

alban reviewed (commented):
Great to see progress on BTFGen :) 

I see it adds 2min 30s in the GHA:
- 2min to git-clone btfhub
- 30s to generate the BTF files

Maybe `--depth 1` helps here.

`pkg/gadgets/socket-collector/tracer/*.o` are modified

mauriciovasquezbernal commented on tools/btfgen.sh:
```diff
@@ -0,0 +1,46 @@
+        --object ${LIBBPFTOOLS}/mountsnoop.bpf.o                    \
+        --object ${LIBBPFTOOLS}/opensnoop.bpf.o                     \
+        --object ${LIBBPFTOOLS}/tcpconnect.bpf.o                    \
+        --object ./pkg/gadgets/seccomp/tracer/bpf/seccomp.o         \
```
Yes, that's right. Those don't work with BTFGen because they required native BTF support in the kernel, so it doesn't make sense to include them here.

mauriciovasquezbernal commented on tools/btfgen.sh:
```diff
@@ -0,0 +1,46 @@
+
+#!/bin/bash
```
Will fix, thanks!

mauriciovasquezbernal commented on gadget-container/entrypoint.sh:
```diff
@@ -144,23 +144,31 @@ ARCH=$(uname -m)
+    echo "shipped BTF available. Installed at /boot/vmlinux-$KERNEL"
   else
-    echo "vmlinux not found"
+    echo "shipped BTF not available. Trying BTFHub"
```
I'll apply it manually.

mauriciovasquezbernal commented on tools/getbtfhub.sh:
```diff
@@ -0,0 +1,10 @@
+set -e
+
+# clone BTFHub and extract files
+git clone https://github.com/aquasecurity/btfhub /tmp/btfhub
```
Yes, it helps. I have to use a workaround because it doesn't work directly with git clone -> https://stackoverflow.com/a/43136160.

mauriciovasquezbernal commented:
> Great to see progress on BTFGen :)

Thanks, I'm looking pretty much to have it integrated. 

> I see it adds 2min 30s in the GHA:
> 
>     * 2min to git-clone btfhub

Actually most of the time is used extracting the files. See the following relevant timestamps:

```
20:44:52.8186586Z #20 0.065 Cloning into '/tmp/btfhub'..
20:45:14.7255536Z #20 22.06 HEAD is now at b2cd89d tools/bin: add btfgen (v1) for arm64
20:46:52.2987002Z #20 119.6 + LIBBPFTOOLS=/objs BTFHUB=/tmp/btfhub ./tools/btfgen.s
```
~22 seconds for cloning 
~98 seconds for extracting (perhaps it can be improved to be multithread)

>     * 30s to generate the BTF files

I think multithread could also help :thinking: 

> `pkg/gadgets/socket-collector/tracer/*.o` are modified

Did you miss part of your message here?

alban commented on tools/getbtfhub.sh:
```diff
@@ -0,0 +1,13 @@
+# https://stackoverflow.com/questions/31278902/how-to-shallow-clone-a-specific-commit-with-depth-1
+mkdir /tmp/btfhub
+pushd /tmp/btfhub
+git remote add origin https://github.com/aquasecurity/btfhub/
+git fetch --depth=1 origin b2cd89dbcf1d40f0304e1e78ab429538b17206a0
```
`git init` missing

blanquicet reviewed (commented):
I think it is great to start integrating BTFGen!! 🎉

The PR looks great, my only concern is the development/testing cycle but I am sure we can find a way to increase it only by the strictly necessary time.

To test it locally I modified the `Makefile` like this, is it the simplest way?
```
docker build -t $(CONTAINER_REPO):$(IMAGE_TAG) -f gadget.Dockerfile --build-arg ENABLE_BTFGEN=true .
```

I tested this PR on an AKS underlay cluster with kernel 5.4.0-1062-azure. Here are my results:

- The vmlinux is indeed lighter: 4.2M (Without this PR and using BTFHub) vs 2.4K (With this PR).
- seccomp: OK.
- bindsnoop: OK.
- execsnoop: OK.
- opensnoop: OK.
- mountsnoop: KO but is also failing without this PR and using BTFHub:
```
$ ./kubectl-gadget mountsnoop -A
/bin/gadgets/mountsnoop: unrecognized option '--mntnsmap'
Try `mountsnoop --help' or `mountsnoop --usage' for more information.
```
- tcpconnect: KO but is also failing without this PR and using BTFHub:
```
libbpf: load bpf program failed: Permission denied
libbpf: -- BEGIN DUMP LOG ---
...
libbpf: -- END LOG --
libbpf: failed to load program 'tcp_v4_connect_ret'
libbpf: failed to load object 'tcpconnect_bpf'
libbpf: failed to load BPF skeleton 'tcpconnect_bpf': -4007
failed to load BPF object: -4007
```

mauriciovasquezbernal commented on gadget.Dockerfile:
```diff
@@ -32,11 +51,10 @@ FROM docker.io/kinvolk/traceloop:20211109004128958575 as traceloop
 # See BCC section in docs/CONTRIBUTING.md for further details.
 
-FROM quay.io/kinvolk/bcc:a81b62e2d04fbebfbbd75787a735edede4555cf8-focal-release
+FROM ${BCC_REPO}:${BCC_TAG} as bcc
```
Not specific reason. Will update.

mauriciovasquezbernal commented on gadget.Dockerfile:
```diff
@@ -20,6 +29,16 @@ RUN cd /gadget && go mod download
 COPY ./ /gadget
 RUN cd /gadget/gadget-container && make gadget-container-deps
 
+# BTFGen
+
+COPY --from=objs /objs /objs
+RUN set -ex; mkdir -p /tmp/btfs && \
+	if [ "$ENABLE_BTFGEN" = true ]; then \
+		cd /gadget && \
+		./tools/getbtfhub.sh && \
+		LIBBPFTOOLS=/objs BTFHUB=/tmp/btfhub ./tools/btfgen.sh; \
+	fi
+
```
Yeah, I think it'll help when building the think locally, however I'm not sure if this will mess up with the Github Actions cache as [it's limit is 5GB](https://docs.github.com/en/actions/advanced-guides/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy) but we're using more than it. :thinking:

mauriciovasquezbernal commented:
> docker build -t $(CONTAINER_REPO):$(IMAGE_TAG) -f gadget.Dockerfile --build-arg ENABLE_BTFGEN=true .

That's right. I extended the makefile to have a `ENABLE_BTFGEN` option. 

> mountsnoop: KO but is also failing without this PR and using BTFHub:

I think it only happens with this PR. It's because I was missing some commits on the branch I used for https://github.com/kinvolk/bcc/pull/15. Once that PR is merged and the BCC repo URL is updated that should work fine.

mauriciovasquezbernal commented:
I think this PR is ready for final review. I don't think there is any documentation we should wrote, the building of the container continues to work the same as before so developers won't have any change.

alban commented on tools/getbtfhub.sh:
```diff
@@ -0,0 +1,13 @@
+# git clone --depth and -b can't be used together. Let's use a workaround
+# https://stackoverflow.com/questions/31278902/how-to-shallow-clone-a-specific-commit-with-depth-1
+
+git clone --depth 1 https://github.com/aquasecurity/btfhub/ /tmp/btfhub
```
URL will need to be updated (see https://github.com/kinvolk/inspektor-gadget/pull/406)

mauriciovasquezbernal replied:
Actually not only the URL but much of the script as the tools are now shipped in other repository. Anyway, it's already done.

mauriciovasquezbernal commented:
I updated the follow the latest changes in BTFHub. BTHHub is more stable now so hopefully we won't hit many more problems with this. 

This PR increases the build time by about 3 minutes, it's not great but we'll look for solutions later on.

blanquicet reviewed (commented):
I tested the involved gadgets in an AKS underlay cluster with kernel 5.4.0-1062-azure and they worked as expected. I didn't tried tcpconnect because the CO-RE version requires kernel 5.8.

I noticed we are now using the [btfhub script](https://github.com/aquasecurity/btfhub/blob/main/tools/btfgen.sh) which extracts the archives and runs btfgen. The only problem I see with that is that we will extract the BTFHub archives every time we modify the code during the development cycle, is it correct? Anyway, as you already said, we can work on optimizations later on.

On the other hand:

> I don't think there is any documentation we should write
>
What about roughly explaining what we are doing now in the [requirements doc](https://github.com/kinvolk/inspektor-gadget/blob/eb3103a8e41cb567aa33f7efe05a0d22db6e5bde/docs/requirements.md) given that we are already talking about BTFHub there?

mauriciovasquezbernal commented on tools/getbtfhub.sh:
```diff
@@ -0,0 +1,12 @@
+set -e
+
+# clone BTFHub and extract files
+# git clone --depth and -b can't be used together. Let's use a workaround
+# https://stackoverflow.com/questions/31278902/how-to-shallow-clone-a-specific-commit-with-depth-1
```
It's not!

mauriciovasquezbernal commented on tools/getbtfhub.sh:
```diff
@@ -0,0 +1,12 @@
+# git clone --depth and -b can't be used together. Let's use a workaround
+# https://stackoverflow.com/questions/31278902/how-to-shallow-clone-a-specific-commit-with-depth-1
+
+git clone --depth 1 https://github.com/aquasecurity/btfhub /tmp/btfhub
+git clone --depth 1 https://github.com/aquasecurity/btfhub-archive/ /tmp/btfhub-archive/
```
I don't think it makes sense. btfhub-archive is used just as a storage solution for the BTF files, having this as a submodule would make the workflow more difficult as the btfhub repo will have to be updated each time btfhub-archive is updated...

blanquicet replied:
Yeah, I was expecting that answer but I still think that it does not make sense to clone both repos separately, if one clones the btfhub repo, 99,99% of the times, one will also clone the btfhub-archive repo. So, I would expect to have a mechanism to make both things in one shot. Anyway, that's completely out of scope for this PR, just sharing my thoughts 🙂.

mauriciovasquezbernal commented on gadget.Dockerfile:
```diff
@@ -12,6 +21,14 @@ RUN set -ex; \
 	apt-get install -y libbpf-dev && \
 	ln -s /usr/lib/go-1.16/bin/go /bin/go
 
+# Download and extract BTFHub files
+COPY ./tools /btf-tools
+RUN set -ex; mkdir -p /tmp/btfs && \
+	if [ "$ENABLE_BTFGEN" = true ]; then \
+		cd /btf-tools && \
+		./getbtfhub.sh; \
+	fi
+
```
> Is there a specific reason why you put this before the go mod download and not after

There is not a strong reason, I think the go modules can change more than btfhub, hence it's better to cache the latter first. However I'm not 100% sure about it. 

> If the developer wants to disable the btfgen after having compiled once with ENABLE_BTFGEN=true

Is that a common case?

blanquicet replied:
> There is not a strong reason, I think the go modules can change more than btfhub, hence it's better to cache the latter first. However I'm not 100% sure about it.
>
Makes sense.

> Is that a common case?
> 
Yeah, I don't think so 😅 .

mauriciovasquezbernal commented on gadget.Dockerfile:
```diff
@@ -1,7 +1,16 @@
+FROM ${BCC} as bcc
 FROM ubuntu:${OS_TAG} as builder
 
+ARG ENABLE_BTFGEN=false
+ENV ENABLE_BTFGEN=${ENABLE_BTFGEN}
```
Good point. I'm still not sure if we should enable it by default. I think most of the developers don't care about it and it adds some extra time to the container building process so we can have it disabled by default in the Makefile too. Anyway, I'll add some documentation about it so people is aware of what's going on.

blanquicet replied:
A great improvement we could think about is to find a way to trigger this process only when the eBPF source files involved (CO-RE BCC libbpf-tools and our seccomp) change. WDYT?

Anyway, we can manage it in a future PR.

mauriciovasquezbernal replied:
Yes, we should find a way to run it only when needed.

mauriciovasquezbernal commented:
> What about roughly explaining what we are doing now in the requirements doc given that we are already talking about BTFHub there?

I changed the sentence we have there and added another one to the CONTRIBUTING document.

blanquicet reviewed (approved):
LGTM!

I think it makes sense to keep it disabled in the Makefile/Dockerfile for now but enabled in the CI. Then, as I said in the comments, we could think about how to trigger the process only when needed (eBPF source code involved changes).

mauriciovasquezbernal commented:
Thanks for the review. I'll open some issues with the possible improvements once it's merged. I'll let it open for another couple of days in case anybody else has comments.
